packages/react/src/components/form/ui-state-machine.ts

```diff
@@ -0,0 +1,343 @@
+          });
+      },
+      exit: () => {
+        this.sid.unsubscribe("otpCodeSent", otpCodeSentHandler);
+      },
+    });
+  }
+
+  send(e: Event) {
+    switch (e.type) {
+      case "sid_ui.otpCodeSent":
+        this.transition(
+          createInputState("input", (otp) => {
+            this.sid.publish("otpCodeSubmitted", otp);
+            this.send({ type: "sid_ui.otpCodeSubmitted" });
+          })
+        );
+        break;
+
+      case "sid_ui.otpCodeSubmitted":
+        this.transition({
+          status: "submitting",
+        });
+        break;
+
+      default:
+        super.send(e);
+    }
+  }
+}
+
+class PasswordUIStateMachine extends UIStateMachine<PasswordStatus> {
+  constructor(opts: UIStateMachineOpts) {
+    const onSetPassword = () => {
+      this.send({ type: "sid_ui.passwordSetReady" });
+    };
+
+    const onVerifyPassword = () => {
+      this.send({ type: "sid_ui.passwordVerifyReady" });
+    };
+
+    super(opts, {
+      status: "initial",
+      entry: () => {
+        this.sid.subscribe("passwordSetReady", onSetPassword);
+        this.sid.subscribe("passwordVerifyReady", onVerifyPassword);
+
+        this.logInFn(this.context.config, this.context.options)
+          .then((user) => {
+            if (user) {
+              this.send({ type: "sid_login.success", user });
+            } else {
+              this.send({
+                type: "sid_login.error",
+                error: new Error("User not returned from /id"),
+              });
+            }
+          })
+          .catch((error) => {
+            this.send({ type: "sid_login.error", error });
+          });
+      },
+      exit: () => {
+        this.sid.unsubscribe("passwordSetReady", onSetPassword);
+        this.sid.unsubscribe("passwordVerifyReady", onVerifyPassword);
+      },
+    });
+  }
+  send(e: Event) {
+    switch ((e as UIEvent).type) {
+      case "sid_ui.passwordSetReady":
+        this.transition(
+          createInputState("setPassword", (password) => {
+            this.sid.publish("passwordSubmitted", password);
+            this.send({ type: "sid_ui.passwordSubmitted" });
+          })
+        );
+        break;
+
+      case "sid_ui.passwordVerifyReady":
+      case "sid_ui.passwordRecovered":
+        this.transition(
+          createVerifyPasswordState(
+            (password) => {
+              this.sid.publish("passwordSubmitted", password);
+              this.send({ type: "sid_ui.passwordSubmitted" });
+            },
+            () => {
+              this.send({ type: "sid_ui.passwordRecoveryRequested" });
+            }
+          )
+        );
+
+        break;
+
+      case "sid_ui.passwordRecoveryRequested":
+        this.transition({
+          status: "recoverPassword",
+          entry: async () => {
+            try {
+              await this.recover();
+              this.send({ type: "sid_ui.passwordRecovered" });
+            } catch (e) {
+              // ignored
```

Comment:
Is this error caught somewhere else? In the existing flow, failing to recover a password should send you to the error state:
```
catch (error) {
      send({ type: "sid_login.error", error: ensureError(error) });
    }
```

Reply:
This was copy-pasted from other place I'm sure, I'll check the behaviour when I move on to QA in another PR